### PR TITLE
[WIP]Paddle fleet distributed strategy

### PR DESCRIPTION
**Pull request by guru4elephant**
### PR types

New features

### PR changes

APIs

### Describe

unify distributed_strategy in paddle.fleet. All distributed training strategy should be configured through this class.

Propose to formalize the paddle.fleet api as follows:

- base
  - distributed_strategy.py
     - all distributed strategies are implemented here, including various kinds of distributed optimizers
     - the distributed strategy can be serialized into protobuf file 
     - distributed job info is defined in distributed_strategy.py, including distributed runtime information such as endpoints, node num, etc.
     - a distributed training job can be configured through serialized distributed strategy and distributed job info.
  - fleet_base.py
     - the only entry of distributed training of Paddle
     - singleton instance 
     - a Fleet() object is initialized through RoleMaker
     - Util and DistributedOptimizer(either collective、ps or other distributed optimizer) can be initialized by RoleMaker
  - obj_creator.py
     - obj_creator helps to create Util and DistributedOptimizer 
  - role_maker.py
     - PaddleCloudRoleMaker and UserDefinedRoleMaker can be defined currently
     - different kinds of RoleMaker is specified by arguments of `__init__(self, role_maker)`
  - util_base.py
     - implement all Util classes, Util Base class can be inherited so that a third party Util class can be used.
- collective
  - collective_distributed_optimizer.py
    - define collective training
- parameter_server
  - ps_distributed_optimizer.py
    - define ps training 
- metrics
  - meric.py
    - distributed metrics

### What is paddle.fleet?
  - **paddle.fleet** is the official entry point for distributed training of Paddle. The motivation of **paddle.fleet** is to **make distributed training efficient and easy to configure**.
  - **RoleMaker, DistributedStrategy and DistributedOptimizer** are the key concepts to understand in **paddle.fleet**
    - **RoleMaker** : the role of a node in distributed cluster is specified by RoleMaker, such as worker or pserver.
    - **DistributedStrategy** : **paddle.fleet** supports lots of distributed training algorithm and scalable configurations all defined in **DistributedStrategy**, such as hierarchical all reduce, deep gradient compression and async parameter server training, etc.
    - **DistributedOptimizer** : in general, three kinds of optimizers exist in paddle. **1)** commonly used local optimizer, such as SGD and Adam. **2)** decorator of local optimzier, such RecomputeOptimizer, PipelineOptimizer and LookAheadOptimizer that often wraps a local optimizer as an inner optimizer. **3)** distributed decorator of **1)、2)** or combination of **1)** and different **2)**, such as collective distributed optimizer and parameter server distributed optimzier. paddle.fleet is the third kind of optimizer that users can define.

### CODE REVIEW

**Review by gongweibao — COMMENTED** [review]
1.代码结构整理和DistributedStrategy的升级能否分成两个PR？现在合在一起，看不到diff。
2.代码结构整理能否用`git mv`，现在的代码和之前的看不到和base的关系!

**Comment by guru4elephant** [comment]
> 1.代码结构整理和DistributedStrategy的升级能否分成两个PR？现在合在一起，看不到diff。
> 2.代码结构整理能否用`git mv`，现在的代码和之前的看不到和base的关系!

本次接口升级可以为不兼容升级，原始的代码目录仍然保留但不能mv
dataset的部分和几个distributed_optimizer已经删掉

File: paddle/fluid/framework/distributed_strategy.proto [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  optional bool localsgd = 6 [ default = false ];
+  optional int32 localsgd_k_step = 7 [ default = 4 ];
+  optional bool dgc = 8 [ default = false ];
+  optional bool hierachical_allreduce = 9 [ default = false ];
+  optional int32 nccl_comm_num = 10 [ default = 1 ];
+  optional bool gradient_merge = 11 [ default = false ];
+  optional int32 gradient_merge_k_step = 12 [ default = 1 ];
+  optional bool sequential_execution = 13 [ default = false ];
+  optional bool enable_backward_optimizer_op_deps = 14 [ default = true ];
+  optional bool elastic = 15 [ default = false ];
```
Comment (gongweibao) [inline paddle/fluid/framework/distributed_strategy.proto]: elastic=401?
感觉elastic和周围的选项是垂直关系。

Comment (guru4elephant) [reply paddle/fluid/framework/distributed_strategy.proto]: 可以修改

**Review by guru4elephant — COMMENTED** [review]
fixed

**Pull request closed without merging**